CDAP-13447 when there are no files matching the time range, write empty files fo…

gsps1 opened the pull request:
…r count, summary and surface results appropriately

wanted to get some feedback on this before wrapping up with test cases.

Context 

Currently if the query time range doesn't match any files
Eg : query end time is before the the first program run in this namespace, so no files are matched. 

Then the reporting job (spark-sql) throws the exception while trying to load the avro files as its empty.

There are two ways to handle this.
1) Reporting app respond with 400 with error message, when no files matched.
2) write empty summary,count files and respond with empty report as result and respond 200 success.

In this PR i took the 2nd approach where i write count file with 0 as count and summary file.

gsps1 committed:
when there are no files matching the time range, write empty files for count, summary and surface results appropriately

poornachandra reviewed (commented):
@gsps1 The approach and the changes look good. Please continue adding the tests.

Can you confirm that the empty files that get written are removed after a certain amount of time?

gsps1 commented:
@poornachandra we are writing _SUCCESS file which has the timestamp when report was successfully completed, which will expire 2 days from that time,  the expired reports are filtered from returning , however currently there is cleaned up after expiration and there is an open JIRA to improve that.

gsps1 committed:
adding test case and covering a case when files are matched but no records are matched to obtain namespaces from request

gsps1 commented:
build : https://builds.cask.co/browse/CDAP-RUT1568-1

albertshau commented on cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java:
```diff
@@ -682,38 +682,42 @@ public void getReportDetails(HttpServiceRequest request, HttpServiceResponder re
+      String total =
+        new String(ByteStreams.toByteArray(reportIdDir.append(LocationName.COUNT_FILE).getInputStream()),
+                   StandardCharsets.UTF_8);
+      Long totalRecords = Long.parseLong(total);
```
long instead of Long

albertshau commented on cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/scala/co/cask/cdap/report/ReportGenerationHelper.scala:
```diff
@@ -51,6 +51,7 @@ object ReportGenerationHelper {
   // the default name of the column created by calling aggregate function count
   val COUNT_COL = "count"
 
+
```
remove extra newline

albertshau commented on cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/scala/co/cask/cdap/report/ReportGenerationHelper.scala:
```diff
@@ -95,6 +96,10 @@ object ReportGenerationHelper {
   @throws(classOf[IOException])
   def generateReport(sql: SQLContext, request: ReportGenerationRequest, inputURIs: java.util.List[String],
                      reportIdDir: Location): Unit = {
+    if  (inputURIs.isEmpty) {
```
remove extra space

albertshau commented on cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/scala/co/cask/cdap/report/ReportGenerationHelper.scala:
```diff
@@ -140,6 +145,47 @@ object ReportGenerationHelper {
+    val namespacesAggregates = getNamespaceAggregates(request)
+
+    val summary = new ReportSummary(namespacesAggregates, request.getStart, request.getEnd,
+      new ArrayBuffer[ArtifactAggregate](), new DurationStats(0l, 0l, 0.0),
```
can you use L instead of l, these look like 01

albertshau commented on cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/scala/co/cask/cdap/report/ReportGenerationHelper.scala:
```diff
@@ -140,6 +145,47 @@ object ReportGenerationHelper {
+    writeSummaryToFile(summary, reportIdDir)
+    val count = 0
+    // Create a _COUNT file and write the total number of report records in it
+    writeToFile(count.toString, Constants.LocationName.COUNT_FILE, reportIdDir)
```
why do we need three separate files instead of a single file? Can't you just put the count and time in the summary?

gsps1 committed:
address comments

gsps1 commented:
@albertshau @poornachandra addressed the comment and now writing the summary, count and expiration time to a single file and added relevant changes in handler code.

albertshau reviewed (approved):
lgtm

gsps1 commented:
build running https://builds.cask.co/browse/CDAP-RUT1568-3

gsps1 commented:
build https://builds.cask.co/browse/CDAP-RUT1568-4 successful, merging.
